### PR TITLE
solving: Default to LU + MUMPS where possible

### PR DESCRIPTION
**Pull request by wence-**
This makes the default user experience friendlier.

### CODE REVIEW

File: firedrake/solving_utils.py [hunk 1]
```diff
@@ -21,6 +23,41 @@ def _make_reasons(reasons):
...
 
 
+if get_config()["options"]["petsc_int_type"] == "int32":
+    DEFAULT_KSP_PARAMETERS = {"mat_type": "aij",
+                              "ksp_type": "preonly",
+                              "pc_type": "lu",
+                              "pc_factor_mat_solver_type": "mumps"}
+else:
+    # FIXME: which factorisation package to use in int64?
+    DEFAULT_KSP_PARAMETERS = {}
```
Comment (pefarrell) [inline firedrake/solving_utils.py]: The MUMPS manual suggests it can be built in fully 64-bit mode, and in "partial" 64-bit mode (where only things that might be big are 64-bit integers). Does PETSc not support it?

File: firedrake/solving_utils.py [hunk 2]
```diff
@@ -21,6 +23,41 @@ def _make_reasons(reasons):
...
+    """Set defaults for solver parameters.
+
+    :arg solver_parameters: dict of user solver parameters to override/extend defaults
+    :arg arguments: arguments for the bilinear form (need to know if we have a Real block).
+    :arg defaults: Any default sets of parameters to supply."""
+    parameters = dict(chain.from_iterable(d.items() for d in defaults))
+    if any(isinstance(V, fsi.RealFunctionSpace)
+           for V in chain.from_iterable(a.function_space() for a in arguments)):
+        # There's a Real block, override with defaults with nest + jacobi.
+        parameters.update({"pc_type": "jacobi", "mat_type": "nest"})
```
Comment (pefarrell) [inline firedrake/solving_utils.py]: I think in this case we should warn the user and generate a set of options that takes a full Schur fieldsplit onto the R-blocks and uses LU for the remainder (à la the Moore-Spence tests in firedrake). This is the most robust solver we can automatically pick, in the absence of assembled R blocks.

File: firedrake/solving_utils.py [hunk 3]
```diff
@@ -21,6 +23,41 @@ def _make_reasons(reasons):
...
+    solver_parameters = solver_parameters or {}
+    mat_type = solver_parameters.get("mat_type")
+    pmat_type = solver_parameters.get("pmat_type")
+    # Allow anything, interpret "matfree" as matrix_free.
+    matfree = mat_type == "matfree"
+    pmatfree = pmat_type == "matfree"
+    # No preconditioner by default for matrix-free
+    if matfree or pmatfree:
+        parameters.update({"pc_type": "none"})
```
Comment (pefarrell) [inline firedrake/solving_utils.py]: Should we automatically set `ksp_type: "fgmres"` and `"ksp_monitor_true_residual": None`, too? It will default to `preonly/none` if the user just passes `{"mat_type": "matfree"}`, which probably is undesirable.

File: firedrake/variational_solver.py [hunk 4]
```diff
@@ -95,7 +95,18 @@ def dm(self):
 class NonlinearVariationalSolver(OptionsManager):
     r"""Solves a :class:`NonlinearVariationalProblem`."""
 
-    def __init__(self, problem, **kwargs):
+    DEFAULT_SNES_PARAMETERS = {"snes_type": "newtonls"}
```
Comment (pefarrell) [inline firedrake/variational_solver.py]: I think we should also default to `"snes_monitor": None` (in case it's not the default) and `"snes_linesearch_type": "basic"` (for compatibility with FEniCS; the PETSc `"bt"` default breaks lots of codes that otherwise would work).

File: firedrake/solving_utils.py (hunk 1 above)
Comment (wence-) [reply firedrake/solving_utils.py]: PETSc buildsystem says:
```python
    # Mumps does NOT work with 64 bit integers without a huge number of hacks we ain't making
    self.precisions       = ['single','double']
    self.requires32bitint = 1;  # 1 means that the package will not work with 64 bit integers
```

File: firedrake/solving_utils.py (hunk 3 above)
Comment (wence-) [reply firedrake/solving_utils.py]: True. I realised when doing this in general that setting up defaults for all the different cases is somewhat fiddly. I'm inclined to do:

1. No options, no real blocks -> LU + preonly
2. No options, real blocks -> either Jacobi or auto-generate fieldsplit options + LU
3. Anything else: no defaults, user knows what they're doing.

File: firedrake/variational_solver.py (hunk 4 above)
Comment (wence-) [reply firedrake/variational_solver.py]: I'm happy to add the linesearch change. The monitoring is a more controversial one: PETSc doesn't turn on monitors by default, although FEniCS does, Firedrake presently doesn't, and this may annoy existing users. Additionally defaulting to `"snes_monitor": None` has the unfortunate side-effect that users will struggle to turn it off: they need to remove it from the dictionary somehow, but if we're shoving it in via defaults...

File: firedrake/solving_utils.py (hunk 1 above)
Comment (pefarrell) [reply firedrake/solving_utils.py]: Fair enough. I guess the next best choice would be `superlu_dist`, but that would require changing the default PETSc configure options too.

File: firedrake/solving_utils.py (hunk 3 above)
Comment (pefarrell) [reply firedrake/solving_utils.py]: I agree (with choice fieldsplit + LU for point 2).

File: firedrake/variational_solver.py (hunk 4 above)
Comment (pefarrell) [reply firedrake/variational_solver.py]: I think they can turn it off with `"snes_monitor_cancel": None`?

Comment (pefarrell) [reply firedrake/variational_solver.py]: Oh, that only works if the monitor has been set up in code, not via options. (Maybe that's the way to go, then.)

File: firedrake/solving_utils.py (hunk 3 above)
Comment (dham) [reply firedrake/solving_utils.py]: Can we detect the fieldsplit reliably enough? E.g. how do we determine which block to eliminate?

Comment (dham) [reply firedrake/solving_utils.py]: If you have a mixed system then even in 2D LU can get big fast and/or defeat the MUMPS memory management heuristics. Is this really a robust option?

**Comment by colinjcotter** [comment]
Also isn't fieldsplit plus LU on the blocks only marginally better than unpreconditioned GMRES for e.g. mixed Poisson?

This feels like a fool's errand ...

________________________________

**Comment by wence-** [comment]
> Also isn't fieldsplit plus LU on the blocks only marginally better than unpreconditioned GMRES for e.g. mixed Poisson? This feels like a fool's errand ...

The proposal is:

1. I can assemble an AIJ matrix -> use LU

2. I can't assemble an AIJ matrix (Real constraints) -> use fieldsplit + LU on the bit I can assemble + sherman-morrison.

File: firedrake/solving_utils.py (hunk 3 above)
Comment (wence-) [reply firedrake/solving_utils.py]: I'm happy to add superlu_dist to the default set of installed packages (it knows about malloc!). In terms of "big", 2D problems with O(1-10e6) dofs are quite happily solved in reasonable time using MUMPs on a workstation.

In terms of defaults, note that in all our tutorials we explicitly say "use LU until you have your numerics/model right and then think about preconditioning". But then we don't default to that option, so you have to learn a load of magic incantations first.

**Comment by pefarrell** [comment]
Yes, if you have n real variables (n small, typically n = 1) then full Schur fieldsplit + LU for (1, 1) block + n iterations of GMRES on the Schur complement solves the problem exactly (in exact arithmetic).

**Comment by pefarrell** [comment]
> If you have a mixed system then even in 2D LU can get big fast and/or defeat the MUMPS memory management heuristics. Is this really a robust option?

David makes a good point. I would suggest also defaulting to `"mat_mumps_icntl_14": 200` (or similar).

**Comment by rckirby** [comment]
I also have frequent headaches trying to use mumps.  Maybe we dodge then if
we we can use the default LU (maybe with a special parameter to keep it
from breaking on saddle points) in serial processor mode -- that's usually
not terrible.

Is there any experience with pastix?  Lawrence and I used it in a paper to
some effect, but I'm not sure how robust it is across a range of problems.

**Comment by wence-** [comment]
OK, so which of these best describes the conclusion of these discussions?

1. Yes, do this (with the better mumps options + fieldsplit/LU for problems with real blocks)
2. Yes, do this, but use superlu_dist instead of mumps (malloc! 64bit!)
3. No, continue with the current state of play.

**Comment by colinjcotter** [comment]
I think that for mixed problems we shouldn’t try to provide an LU default (except for R).

On 10 Jul 2019, at 16:58, Lawrence Mitchell <notifications@github.com<mailto:notifications@github.com>> wrote:


OK, so which of these best describes the conclusion of these discussions?

  1.  Yes, do this (with the better mumps options + fieldsplit/LU for problems with real blocks)
  2.  Yes, do this, but use superlu_dist instead of mumps (malloc! 64bit!)
  3.  No, continue with the current state of play.

**Comment by pefarrell** [comment]
I vote for option #1. With `"mat_mumps_icntl_14": 200` MUMPS is almost bulletproof.

@colinjcotter No one is advocating fieldsplits when not necessary (they are necessary with real blocks because of technical limitations).

**Comment by rckirby** [comment]
If I don’t have to remember that mumps option and it’s robust, it’s fine with me.

File: firedrake/solving_utils.py (hunk 1 above)
Comment (wence-) [reply firedrake/solving_utils.py]: #1474 changes petsc defaults and installs superlu_dist by default, so if that lands, I will do that here.

File: firedrake/solving_utils.py (hunk 3 above)
Comment (wence-) [reply firedrake/solving_utils.py]: I've changed to (hopefully) do the 3-part proposal above (with fieldsplit + Schur elimination for problems with Real blocks).

File: firedrake/solving_utils.py [hunk 5]
```diff
@@ -21,6 +24,78 @@ def _make_reasons(reasons):
...
+                    "fieldsplit_0": {
+                        "ksp_type": "preonly",
+                        "pc_type": "python",
+                        "pc_python_type": "firedrake.AssembledPC",
+                        "assembled": DEFAULT_KSP_PARAMETERS,
+                    },
+                    "fieldsplit_1": {
+                        "ksp_type": "gmres",
+                        "pc_type": "jacobi",
+                    }}
```
Comment (wence-) [inline firedrake/solving_utils.py]: @pefarrell does this look right?

File: firedrake/variational_solver.py [hunk 6]
```diff
@@ -95,7 +95,19 @@ def dm(self):
 class NonlinearVariationalSolver(OptionsManager):
     r"""Solves a :class:`NonlinearVariationalProblem`."""
 
-    def __init__(self, problem, **kwargs):
+    DEFAULT_SNES_PARAMETERS = {"snes_type": "newtonls",
+                               "snes_linesearch_type": "basic"}
```
Comment (wence-) [inline firedrake/variational_solver.py]: Added basic linesearch type.

**Comment by wence-** [comment]
If we're happy with the proposal of options, I think this is now right. If not, I'll leave it until agreement is reached on what we want.

File: firedrake/solving_utils.py (hunk 5 above)
Comment (pefarrell) [reply firedrake/solving_utils.py]: No, it wasn't quite right, for two reasons:

a) The catch for checking if any function spaces involved were R0 was wrong (wasn't tripping even when R0 was present)
b) Jacobi needs `MatGetDiagonal`, which isn't implemented.

I changed the condition to look at the element family, which now works, and changed `"jacobi"` for `"none"`.

I've changed `tests/regression/test_moore_spence.py` (which solves an augmented system of PDEs for a bifurcation point, which is a real value) to use the default solver parameters, and it now passes.

**Comment by pefarrell** [comment]
It would be good to have a test that just solved a problem with R-blocks.

**Comment by wence-** [comment]
I think I finally fixed all the corner cases. Let's hope tests pass.

**Comment by wence-** [comment]
Ready for review. All tests are passing now.

File: firedrake/variational_solver.py [hunk 7]
```diff
@@ -321,30 +316,9 @@ def __init__(self, a, L, u, bcs=None, aP=None,
...
-               space of the operator.
-        :kwarg transpose_nullspace: as for the nullspace, but used to
-               make the right hand side consistent.
-        :kwarg options_prefix: an optional prefix used to distinguish
-               PETSc options.  If not provided a unique prefix will be
-               created.  Use this option if you want to pass options
-               to the solver from the command line in addition to
-               through the ``solver_parameters`` dict.
-        :kwarg appctx: A dictionary containing application context that
-               is passed to the preconditioner if matrix-free.
```
Comment (wence-) [inline firedrake/variational_solver.py:340]: Reinstate docstring part of this as class docstring.

File: firedrake/solving_utils.py [hunk 8]
```diff
@@ -21,6 +23,101 @@ def _make_reasons(reasons):
...
+                              "pc_type": "lu",
+                              "pc_factor_mat_solver_type": "mumps",
+                              "mat_mumps_icntl_14": 200}
+else:
+    DEFAULT_KSP_PARAMETERS = {"mat_type": "aij",
+                              "ksp_type": "preonly",
+                              "ksp_rtol": 1e-7,
+                              "pc_type": "lu",
+                              "pc_factor_mat_solver_type": "superlu_dist"}
+
```
Comment (ksagiyam) [inline firedrake/solving_utils.py:39]: I remember we discussed this in the meeting, but I forget. Why do we use `superlu` here instead of `mumps`?

Comment (wence-) [reply firedrake/solving_utils.py:39]: mumps doesn't have a 64bit mode.
